tremor-pipeline/src/op/qos/wal.rs

```diff
@@ -184,47 +196,60 @@ op!(WalFactory(_uid, node) {
     if let Some(map) = &node.config {
         let config: Config = Config::new(map)?;
 
-        Ok(Box::new(Wal::new(node.id.to_string(), config)?))
+        if let (None, None) = (config.max_elements, config.max_bytes) {
+            Err(ErrorKind::BadOpConfig("WAL operator needs at least one of `max_elements` or `max_bytes` config entries.".to_owned()).into())
+        } else {
+            Ok(Box::new(Wal::new(node.id.to_string(), config)?))
+        }
 
     } else {
         Err(ErrorKind::MissingOpConfig(node.id.to_string()).into())
     }
 });
 
 impl Wal {
+    const URI_SCHEME: &'static str = "tremor-wal";
+    const READ: &'static str = "read";
+    const EVENTS: &'static str = "events";
+    const STATE: &'static str = "state";
+
     fn new(id: String, config: Config) -> Result<Self> {
         let wal = if let Some(dir) = &config.dir {
             sled::open(&dir)?
         } else {
             sled::Config::default().temporary(true).open()?
         };
-        let events_tree = wal.open_tree("events")?;
-        let state_tree = wal.open_tree("state")?;
+        let events_tree = wal.open_tree(Self::EVENTS)?;
+        let state_tree = wal.open_tree(Self::STATE)?;
 
         #[allow(clippy::cast_possible_truncation)]
-        let read = state_tree.get("read")?.map(Idx::from).unwrap_or_default();
+        let read = state_tree
+            .get(Self::READ)?
+            .map(Idx::from)
+            .unwrap_or_default();
         Ok(Wal {
             cnt: events_tree.len() as u64,
             wal,
             read,
-            confirmed: read,
+            confirmed: None,
```

Comment:
I think it would make sense to track the last read element at least if we store it. This would allow us to handle GC (deletion) more gracefully down the road.

Reply:
I hope what I added is what you meant.